cli/listener/src/main.rs

```diff
@@ -210,74 +215,417 @@ pub fn get_apy_for_period(
+                }
+                ListenerResult::ErrListener(err) => {
+                    println!("Error in listener.");
+                    err.print_pretty();
+                    self.get_sleep_time_after_error()
+                }
+            };
+            // Update metrics snapshot.
+            *self.metrics_snapshot.lock().unwrap() = Arc::new(self.metrics.clone());
+            std::thread::sleep(sleep_time);
+        }
+    }
+
+    fn get_sleep_time_after_error(&mut self) -> Duration {
+        // For the sleep time we use exponential backoff with jitter [1]. By taking
+        // the time since the last success as the target sleep time, we get
+        // exponential backoff. We clamp this to ensure we don't wait indefinitely.
+        // 1: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+        let time_since_last_success = self.last_read_success.elapsed();
+        let min_sleep_time = Duration::from_secs_f32(0.2);
+        let max_sleep_time = Duration::from_secs_f32(300.0);
+        let target_sleep_time = time_since_last_success.clamp(min_sleep_time, max_sleep_time);
+        let sleep_time = self
+            .rng
+            .gen_range(Duration::from_secs(0)..target_sleep_time);
+        println!("Sleeping {:?} after error ...", sleep_time);
+        sleep_time
+    }
+
+    pub fn get_sleep_time(&self) -> Duration {
+        // Sleep until is time to get the next exchange rate.
+        let sleep_time = std::time::Duration::from_secs(self.opts.poll_frequency_seconds as u64);
+        println!(
+            "Sleeping for {:?} after getting the Solido exchange rate",
+            sleep_time
+        );
+        sleep_time
+    }
+}
+
+#[derive(Clone)]
+struct Metrics {
+    /// Number of times that we checked the price.
+    polls: u64,
+
+    /// Number of times that we tried to get the exchange rate, but encountered an error.
+    errors: u64,
+
+    /// Solido's maximum price interval.
+    solido_average_interval_price: Option<IntervalPrices>,
+}
+
+impl Metrics {
+    pub fn write_prometheus<W: std::io::Write>(&self, out: &mut W) -> std::io::Result<()> {
+        write_metric(
+            out,
+            &MetricFamily {
+                name: "solido_pricedb_polls_total",
+                help: "Number of times we polled the exchange rate, since launch.",
+                type_: "counter",
+                metrics: vec![Metric::new(self.polls)],
+            },
+        )?;
+        write_metric(out, &MetricFamily {
+            name: "solido_pricedb_errors_total",
+            help: "Number of times we encountered an error while trying to get the exchange rate, since launch.",
+            type_: "counter",
+            metrics: vec![Metric::new(self.errors)]
+        })?;
+        if let Some(interval_price) = &self.solido_average_interval_price {
+            write_metric(
+                out,
+                &MetricFamily {
+                    name: "average_apy",
```

Comment:
```suggestion
                    name: "solido_pricedb_average_apy",
```

The name should also have `solido_pricedb_` prefix, because Prometheus doesn’t by itself namespace the metrics; applications have to choose their names to not clash themselves.

Also, we are making the same mistake now that we are building this thing for to avoid :see_no_evil: “average APY” ... over what time period? We should clarify that too, e.g. `solido_pricedb_30d_average_apy`.

But maybe it’s better to not return the `IntervalPrices` from the sampling loop, because the interval we choose is arbitrary. We could just return the current price. And if you have the current price as a metric — but we have that already anyway from the maintainer daemon — you can easily compute the desired APY through e.g. `(solido_pricedb_st_sol_price_in_sol / solido_pricedb_st_sol_price_in_sol offset 30d)^(365/30)`.

Reply:
Great! That's much better, thanks. I just did `now() - 30d` and used the APY function to calculate it. Should be working.